lms/djangoapps/certificates/signals.py

```diff
@@ -68,6 +70,24 @@ def listen_for_passing_grade(sender, user, course_id, **kwargs):  # pylint: disa
     downstream signal from COURSE_GRADE_CHANGED
     """
     course = CourseOverview.get_from_id(course_id)
+    context = contexts.course_context_from_course_id(course_id)
+    data = {
+        'user_id': user.id,
+        'username': user.username,
+        'course_id': str(course_id),
+    }
+    segment_properties = {
+        'category': 'conversion',
+        'label': str(course_id),
+        'org': course_id.org,
+        'course': course_id.course,
+        'run': course_id.run,
+        'email': user.email
+    }
+    event_name = 'edx.course.completed'
+    with tracker.get_tracker().context(event_name, context):
+        tracker.emit(event_name, data)
+        segment.track(user.id, event_name, segment_properties)
     if not auto_certificate_generation_enabled():
         return
```

Comment:
My understanding is that for the purposes of establishing an accurate completion event for all cases, including instructor-led courses which may have auto certification generation _disabled_, we may need this conditional to be placed before the event emission code.  It seems like the closer event emission is to actually generating the certificate, the better?

Reply:
We are going to emit this event course grade passed weather a certificate is enabled or not. So we are considering a course as completed when course grades are passed. Basically we want to add support for course completion event tracking that is supported in edx-enterprise. And in edx-enterprise they as considering a course as completed when we have passed course grande against an enrolled course. That is why we are emitting this event before certificates enabled check. Furthermore, We are glad to here more about course completion exact criteria.

Reply:
What if a learner has previously "completed" ie reached a passing grade, then had their score drop below passing (for example by reattempting a problem?) and then passes again. Would we be firing this event twice? Is that desired behavior?

Reply:
@Nickett3 I expect what you say to be true.  The `COURSE_GRADE_NOW_PASSED` signal is fired any time the CourseGradeFactory produces a passing grade for a given user/courserun, so every time a learner's grade is recalculated and it is passing, it should produce one of these course_completed events.  I also would expect to see multiple consecutive course_completed events if the user continues to interact with the course modules after passing the first time.

I just had a conversation with Jackie and BrianW about event hygiene when it comes to course completion, and we agreed on the following proposal:

* Rename `edx.course.completed` to `edx.course.passing_grade` or something more accurate in general.  We believe that this event, as written, does NOT represent course completion for instructor-led courses.
* Add a complementary `edx.course.failing_grade` event in the receiver of `COURSE_GRADE_NOW_FAILED` on line 98: <https://github.com/edly-io/edx-platform/pull/243/files#diff-5a0fdcbc9440a85ed23d22b8804865bafe7a70e0275d67f6d3c198a12641ec55R98>.  That way, downstream event consumers (particularly, Amplitude users) have a chance at being able to identify passing learners in all possible cases.

What are your thoughts on these recommendations @rehanedly ?

Reply:
On reflection, I like the names of the signals -- to me they are just a little less ambiguous in indicating the transition from fail to pass or pass to fail.  I would suggest a name of edx.course.grade.now_passed and edx.course.grade.now_failed. (Also, the event should have courserun_key used instead of course_id for the tracking log event's data, and add the courserun_key on Segment.)

I know from personal experience this spring that learners can achieve a passing grade in an instructor-led course, and with a change to the grading rubric, return to a non-passing grade, and then later do enough to be able to pass again.   So while it doesn't happen often, it certainly *can* happen, Nick.  And in that case, the *real* completion would be the second one.   Note too that the "now_failed" signal is able to revoke certificates from those in self-paced courses (e.g. with auto-generated certs), so presumably that does happen as well on occasion.   That means that even if we were basing it on when certs are issued, that's also not necessarily final.

Reply:
Thanks all for your suggestions. After detailed discussion we decided to emit 3 events:
1. edx.course.grade.now_passed event from lms/djangoapps/certificates/signals.py on COURSE_GRADE_NOW_PASSED receiver in listen_for_passing_grade function
2. edx.course.grade. now_failed event from lms/djangoapps/certificates/signals.py on COURSE_GRADE_NOW_FAILED receiver in _listen_for_failing_grade function

Reply:
Here is following PR:
https://github.com/edx/edx-platform/pull/28415